src/tree/identity_tree.rs

```diff
@@ -0,0 +1,612 @@
+use std::collections::{BTreeMap, HashMap, VecDeque};
+
+use eyre::{eyre, OptionExt};
+use semaphore::dynamic_merkle_tree::DynamicMerkleTree;
+use semaphore::merkle_tree::{Branch, Hasher};
+use semaphore::poseidon_tree::{PoseidonHash, Proof};
+use semaphore::Field;
+use serde::Serialize;
+
+use super::Hash;
+
+pub enum LeafUpdates {
+    Insert(Leaves),
+    Delete(Leaves),
+}
+
+impl From<LeafUpdates> for Leaves {
+    fn from(val: LeafUpdates) -> Self {
+        match val {
+            LeafUpdates::Insert(leaves) => leaves,
+            LeafUpdates::Delete(leaves) => leaves,
+        }
+    }
+}
+
+// Node index to hash, 0 indexed from the root
+pub type StorageUpdates = HashMap<u32, Hash>;
+
+// Leaf index to hash, 0 indexed from the initial leaf
+pub type Leaves = HashMap<u32, Hash>;
+
+pub fn leaf_to_storage_idx(leaf_idx: u32, tree_depth: usize) -> u32 {
+    let leaf_0 = (1 << tree_depth) - 1;
+    leaf_0 + leaf_idx
+}
+
+pub fn storage_to_leaf_idx(storage_idx: u32, tree_depth: usize) -> u32 {
+    let leaf_0 = (1 << tree_depth) - 1;
+    storage_idx - leaf_0
+}
+
+pub fn storage_idx_to_coords(index: usize) -> (usize, usize) {
+    let depth = (index + 1).ilog2();
+    let offset = index - (2usize.pow(depth) - 1);
+    (depth as usize, offset)
+}
+
+#[derive(PartialEq, Eq, Hash, Clone, Copy, Debug)]
+pub struct Root {
+    pub hash: Hash,
+    //NOTE: note that this assumes that there is only one wallet that sequences transactions
+    // we should update to a syncing mechanism that can account for multiple sequencers
+    pub nonce: usize,
+}
+
+//TODO: comments as to why we only compare nonce and for parial eq we compare hash
+impl Ord for Root {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.nonce.cmp(&other.nonce)
+    }
+}
+
+impl PartialOrd for Root {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+pub struct IdentityTree {
+    pub tree: DynamicMerkleTree<PoseidonHash>,
+    pub tree_updates: BTreeMap<Root, StorageUpdates>,
```

Comment:
Not necessary in this PR but let's remember to add metrics that display stats about these various hash maps (how many entries we have in them). I feel like it'll be easy to forget to remove entries

Reply:
Sounds good, I am planning to open a PR to update logging/metrics after this is merged.